hub/core/storage/prefetch_lru_cache.py

```diff
@@ -0,0 +1,342 @@
+import numpy as np
+from itertools import repeat
+from pathos.pools import ProcessPool  # type: ignore
+from typing import Callable, Dict, Optional, Sequence, Tuple, Union, List
+
+from hub.constants import MB
+from hub.core.chunk import Chunk
+from hub.core.chunk_engine import ChunkEngine
+from hub.core.storage import (
+    S3Provider,
+    LRUCache,
+    StorageProvider,
+    MemoryProvider,
+    SharedMemoryProvider,
+    LocalProvider,
+)
+from hub.util.exceptions import (
+    DatasetUnsupportedPytorch,
+    TensorDoesNotExistError,
+)
+from hub.util.remove_cache import get_base_storage
+from hub.util.prefetch_cache import read_and_store_chunk_groups
+from hub.util.iterable_ordered_dict import IterableOrderedDict
+
+
+class PrefetchLRUCache(LRUCache):
+    """Creates an cache that fetches multiple chunks parallely."""
+
+    def __init__(
+        self,
+        cache_storage: SharedMemoryProvider,
+        next_storage: Optional[StorageProvider],
+        cache_size: int,
+        dataset,
+        num_workers: int,
+        tensor_keys: Optional[Sequence[str]],
+        transform: Callable,
+    ):
+        super().__init__(cache_storage, next_storage, cache_size)
+        global torch
+        import torch
+
+        self.transform = transform
+        self.all_indexes = self._extract_indexes_from_dataset(dataset)
+        self.tensor_keys = self._get_tensor_keys(tensor_keys, dataset)
+        self.workers = num_workers
+        self.map = ProcessPool(nodes=num_workers).map
+
+        # shared memory file names have format "al_{x}" where x is last_shm_key_generated, which is incremented by 1 every time
+        self.last_shm_key_generated = -1
+
+        # keeps track of the last index suggested from all_indexes, incremented by 1 every time to return sequential indexes
+        self.last_index_suggested = -1
+        self.length = len(dataset)
+
+        self.storage = get_base_storage(dataset.storage)
+        if isinstance(self.storage, MemoryProvider):
+            raise DatasetUnsupportedPytorch(
+                "Datasets whose underlying storage is MemoryProvider are not supported for Pytorch iteration."
```

Comment:
```suggestion
                "Datasets whose underlying storage is MemoryProvider are not supported for prefetch caching."
```

Reply:
can we try to remove as much pytorch references / requirements as possible in here?